File: MMDB/MovieListViewController.m
//
//  MovieListViewController.m
//  MMDB
//
//  Created by Rob Wyland on 8/13/15.
//  Copyright (c) 2015 Org Inc. All rights reserved.
//

#import "MovieListViewController.h"
#import "MovieDetailsViewController.h"
#import "MovieDBService.h"
#import "MoviePosterUtil.h"
#import "Movie.h"

#define kHOT_TITLE @"What's Hot!"
#define kSEARCH_TITLE @"Search"

@interface MovieListViewController () <UITableViewDelegate, UITableViewDataSource, UISearchBarDelegate, MovieLookupDelegate>

@property (strong, nonatomic) UITableView *tableView;
@property (strong, nonatomic) NSMutableArray *moviesDatasource;

@property (strong, nonatomic) MovieDBService *movieDBService;

@property (strong, nonatomic) UISearchBar *searchBar;
@property (strong, nonatomic) NSString *query;
@property (nonatomic, assign) BOOL isSearching;

@end

@implementation MovieListViewController

#pragma mark - Init

- (instancetype)init {
    if ( self = [super init] ) {
        [self initialize];
    }
    return self;
}

- (id)initWithCoder:(NSCoder *)aDecoder {
    if ( self = [super initWithCoder:aDecoder] ) {
        [self initialize];
    }
    return self;
}

- (instancetype)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil {
    if ( self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil] ) {
        [self initialize];
    }
    return self;
}

/**
 *  Make sure the private instance variables are correctly initialized, no matter the implementation method
 */
- (void)initialize {
    _tableView = [[UITableView alloc] initWithFrame:CGRectZero style:UITableViewStylePlain];
    _movieDBService = [[MovieDBService alloc] init];
    _moviesDatasource = [NSMutableArray array];
    _searchBar = [[UISearchBar alloc] initWithFrame:CGRectMake(0.f, 0.f, 200.f, 40.f)];
    _isSearching = NO;
}

#pragma mark - View Lifecycle

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self.view addSubview:_tableView];
    [self.tableView setFrame:self.view.bounds];
    [self.tableView setDelegate:self];
    [self.tableView setDataSource:self];
    
    [self.movieDBService setDelegate:self];
    [self.movieDBService requestPopularMoviesWithPage:1];
    
    [self.navigationItem setTitle:kHOT_TITLE];
    UIBarButtonItem *search = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemSearch target:self action:@selector(onSearch:)];
    [self.navigationItem setRightBarButtonItem:search];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    
    [self.tableView reloadData];
}

#pragma mark - TableView Delegate

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    return 150.f;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    
    // TODO: This request can appear to "hang" if the network is slow or even offline. Ideally we want to provide feedback
    Movie *movie = self.moviesDatasource[indexPath.row];
    [self.movieDBService lookupMovieWithId:movie.id];
}

- (void)tableView:(UITableView *)tableView willDisplayCell:(UITableViewCell *)cell forRowAtIndexPath:(NSIndexPath *)indexPath {
    // Once we are getting near the bottom, go ahead and load the next page into the stack
    // If we are on the last page, the next page will return an empty array in the delegate call
    if ( indexPath.row >= self.moviesDatasource.count - 1 ) {
        NSUInteger nextPage = (self.moviesDatasource.count / 20) + 1;
        // TODO: This request can appear to "hang" if the network is slow or even offline. Ideally we want to provide feedback
        self.isSearching ? [self.movieDBService searchMoviesWithQuery:self.query andPage:nextPage] : [self.movieDBService requestPopularMoviesWithPage:nextPage];
    }
}

#pragma mark - TableView Datasource

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.moviesDatasource.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    // Using a simple layout with some slight modifications.
    // This could always use a subclass of UITableCell to use a custom design with more info like movie rating or vote rating
    static NSString *cellIdentifier = @"MovieCell";
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:cellIdentifier];
    if (cell == nil) {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleSubtitle reuseIdentifier:cellIdentifier];
        [cell.textLabel setFont:[UIFont systemFontOfSize:24.f]];
        [cell.textLabel setNumberOfLines:3];
        [cell.detailTextLabel setFont:[UIFont systemFontOfSize:18.f]];
        [cell.detailTextLabel setTextColor:[UIColor colorWithWhite:0.35f alpha:1.f]];
        [cell setAccessoryType:UITableViewCellAccessoryDisclosureIndicator];
    }
    
    Movie *movie = self.moviesDatasource[indexPath.row];
    [cell.textLabel setText:movie.title];
    [cell.detailTextLabel setText:movie.year];
    [MoviePosterUtil loadThumbnailPosterForMovie:movie intoView:cell.imageView];
    
    return cell;
}

#pragma mark - Search Bar Delegate

- (void)searchBarTextDidBeginEditing:(UISearchBar *)searchBar {
    // Add a grey effect and don't allow scrolling, this could also be done with an animation
    [self.tableView setAlpha:0.5f];
    [self.tableView setUserInteractionEnabled:NO];
}

- (void)searchBarTextDidEndEditing:(UISearchBar *)searchBar {
    // remove grey effect and allow scrolling, this could also be done with an animation
    [self.tableView setAlpha:1.f];
    [self.tableView setUserInteractionEnabled:YES];
}

- (void)searchBarSearchButtonClicked:(UISearchBar *)searchBar {
    if ( searchBar.text && searchBar.text.length ) {
        [self searchBarTextDidEndEditing:searchBar];
        self.isSearching = YES;
        self.query = [searchBar.text stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
        [self.movieDBService searchMoviesWithQuery:self.query andPage:1];
    }
}

#pragma mark - Movie DB Delegate

- (void)receivedMovies:(NSArray *)movies forPage:(NSUInteger)page {
    if ( movies.count == 0 && page == 1 ) {
        // The contents are empty, provide feedback to the user
        // We probably dont want to be annoying with an alert, so something like using the table footer would also work
        dispatch_async(dispatch_get_main_queue(), ^{
            UIAlertView *alert = [[UIAlertView alloc] initWithTitle:@"No Results" message:@"please try again" delegate:nil cancelButtonTitle:@"OK" otherButtonTitles:nil];
            [alert show];
        });
        return;
    } else if ( movies.count == 0 ) {
        // Empty page, nothing new to show so do nothing
        return;
    } else if ( page == 1 ) {
        // Start over, new request!
        [self.moviesDatasource removeAllObjects];
    }
    [self.moviesDatasource addObjectsFromArray:movies];
    
    // Always reload the view on the main thread
    dispatch_async(dispatch_get_main_queue(), ^{
        [self.tableView reloadData];
    });
}

- (void)receivedMovieDetails:(Movie *)movie {
    // Always make changes to the view heiarchy on the main thread
    dispatch_async(dispatch_get_main_queue(), ^{
        MovieDetailsViewController *detailsVC = [[MovieDetailsViewController alloc] initWithNibName:@"MovieDetailsViewController" bundle:[NSBundle mainBundle]];
        [detailsVC setMovie:movie];
        [self.navigationController pushViewController:detailsVC animated:YES];
    });
}

- (void)requestFailedWithError:(NSError *)error {
    // An error happened, let the user know!
    // Ideally we would want user friendly messages, with some sort of next-step for the user to try
    dispatch_async(dispatch_get_main_queue(), ^{
        UIAlertView *alert = [[UIAlertView alloc] initWithTitle:@"Network Fail" message:error.description delegate:nil cancelButtonTitle:@"OK" otherButtonTitles:nil];
        [alert show];
    });
}

#pragma mark - Controller Actions

- (void)onSearch:(id)sender {
    // Swap search with a cancel button
    UIBarButtonItem *cancel = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemCancel target:self action:@selector(onCancel:)];
    [self.navigationItem setRightBarButtonItem:cancel animated:YES];

    // This could also have an animation
    [self.navigationItem setTitle:kSEARCH_TITLE];
    [self.navigationItem setTitleView:self.searchBar];
    [self.searchBar setDelegate:self];
    [self.searchBar becomeFirstResponder];
}

- (void)onCancel:(id)sender {
    // Swap cancel with a search button
    UIBarButtonItem *search = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemSearch target:self action:@selector(onSearch:)];
    [self.navigationItem setRightBarButtonItem:search];
    
    // This could also have an animation
    [self.navigationItem setTitle:kHOT_TITLE];
    [self.navigationItem setTitleView:nil];
    [self.searchBar setDelegate:nil];
    
    // Switch back to the popular results
    self.isSearching = NO;
    self.query = nil;
    [self.movieDBService requestPopularMoviesWithPage:1];
}

@end

File: MMDB Swift/MovieListViewController.swift
//
//  MovieListViewController.swift
//  MMDB Swift
//
//  Created by Rob W on 12/11/15.
//  Copyright © 2015 Org Inc. All rights reserved.
//

import UIKit

let kHOT_TITLE: String = "What's Hot!"
let kSEARCH_TITLE: String = "Search"

class MovieListViewController: UIViewController {
  
  var tableView: UITableView = UITableView(frame: .zero, style: .Plain)
  var moviesDatasource: Array<Movie> = []
  
  let movieDBService: MovieDBService = MovieDBService()
  
  let searchBar: UISearchBar = UISearchBar(frame: CGRectMake(0, 0, 200, 40))
  var query: String? = nil
  var isSearching: Bool = false
  
  required init?(coder aDecoder: NSCoder) {
    super.init(coder: aDecoder)
  }
  
  override func viewDidLoad() {
    super.viewDidLoad()
    
    self.view.addSubview(tableView)
    self.tableView.frame = self.view.bounds
    self.tableView.delegate = self
    self.tableView.dataSource = self
    
    self.movieDBService.delegate = self
    self.movieDBService.requestPopularMoviesWithPage(1)
    
    self.navigationItem.title = kHOT_TITLE
    let search = UIBarButtonItem(barButtonSystemItem: .Search, target: self, action: Selector("onSearch:"))
    self.navigationItem.rightBarButtonItem = search
  }
  
  override func viewWillAppear(animated: Bool) {
    super.viewWillAppear(animated)
   
    self.tableView.reloadData()
  }
  
  // MARK: - Controller Actions
  
  func onSearch(sender: AnyObject?) {
    // Swap search with a cancel button
    let cancel: UIBarButtonItem = UIBarButtonItem(barButtonSystemItem: .Cancel, target: self, action: Selector("onCancel:"))
    self.navigationItem.setRightBarButtonItem(cancel, animated: true)
    
    // This could also have an animation
    self.navigationItem.title = kSEARCH_TITLE
    self.navigationItem.titleView = self.searchBar
    self.searchBar.delegate = self
    self.searchBar.becomeFirstResponder()
  }
  
  func onCancel(sender: AnyObject?) {
    // Swap cancel with a search button
    let search: UIBarButtonItem = UIBarButtonItem(barButtonSystemItem: .Search, target: self, action: Selector("onSearch:"))
    self.navigationItem.setRightBarButtonItem(search, animated: true)
    
  
    // This could also have an animation
    self.navigationItem.title = kHOT_TITLE
    self.navigationItem.titleView = nil
    self.searchBar.delegate = nil
  
    // Switch back to the popular results
    self.isSearching = false
    self.query = nil
    self.movieDBService.requestPopularMoviesWithPage(1)
  }
}

// MARK: - Table Delegate and Datasource protocol

extension MovieListViewController: UITableViewDelegate {
  func tableView(tableView: UITableView, heightForRowAtIndexPath indexPath: NSIndexPath) -> CGFloat {
    return 150
  }
  
  func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: NSIndexPath) {
    tableView.deselectRowAtIndexPath(indexPath, animated: true)
    
    // TODO: This request can appear to "hang" if the network is slow or even offline. Ideally we want to provide feedback
    let movie: Movie = self.moviesDatasource[indexPath.row]
    self.movieDBService.lookupMovieWithId(movie.id)
  }
  
  func tableView(tableView: UITableView, willDisplayCell cell: UITableViewCell, forRowAtIndexPath indexPath: NSIndexPath) {
    let movie: Movie = self.moviesDatasource[indexPath.row]
    cell.textLabel!.text = movie.title
    cell.detailTextLabel!.text = movie.year()
    MoviePosterUtil.loadThumbnailPosterForMovie(movie, intoView:cell.imageView!)
    
    if indexPath.row >= self.moviesDatasource.count - 1 {
      let nextPage = (self.moviesDatasource.count / 20) + 1
      
      // TODO: This request can appear to "hang" if the network is slow or even offline. Ideally we want to provide feedback
      self.isSearching ? self.movieDBService.searchMoviesWithQuery(self.query!, andPage:nextPage) : self.movieDBService.requestPopularMoviesWithPage(nextPage)
    }
  }
}

extension MovieListViewController: UITableViewDataSource {
  func numberOfSectionsInTableView(tableView: UITableView) -> Int {
    return 1
  }
  
  func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
    return self.moviesDatasource.count
  }
  
  func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
    // Using a simple layout with some slight modifications.
    // This could always use a subclass of UITableCell to use a custom design with more info like movie rating or vote rating
    let cellIdentifier = "MovieCell"
    guard let cell: UITableViewCell = tableView.dequeueReusableCellWithIdentifier(cellIdentifier) else {
      let newCell = UITableViewCell(style: .Subtitle, reuseIdentifier: cellIdentifier)
      newCell.textLabel!.font = UIFont.systemFontOfSize(24)
      newCell.textLabel!.numberOfLines = 3
      newCell.detailTextLabel!.font = UIFont.systemFontOfSize(18)
      newCell.detailTextLabel!.textColor = UIColor(white: 0.35, alpha: 1)
      newCell.accessoryType = .DisclosureIndicator
      return newCell
    }
    return cell
  }
}

// MARK: - Search Bar Delegate

extension MovieListViewController: UISearchBarDelegate {
  func searchBarTextDidBeginEditing(searchBar: UISearchBar) {
    // Add a grey effect and don't allow scrolling, this could also be done with an animation
    self.tableView.alpha = 0.5
    self.tableView.userInteractionEnabled = false
  }
  
  func searchBarTextDidEndEditing(searchBar: UISearchBar) {
    // remove grey effect and allow scrolling, this could also be done with an animation
    self.tableView.alpha = 1
    self.tableView.userInteractionEnabled = true
  }
  
  func searchBarSearchButtonClicked(searchBar: UISearchBar) {
    if let text = searchBar.text where text.characters.count > 0 {
      searchBarTextDidEndEditing(searchBar)
      self.isSearching = true
      self.query = text.stringByTrimmingCharactersInSet(NSCharacterSet.whitespaceAndNewlineCharacterSet())
      self.movieDBService.searchMoviesWithQuery(self.query!, andPage: 1)
    }
  }
}

// MARK: - Movie Lookup Delegate

extension MovieListViewController: MovieLookupDelegate {
  func receivedMovies(movies: Array<Movie>, forPage page: Int) {
    if movies.count == 0 && page == 1 {
      // The contents are empty, provide feedback to the user
      // We probably dont want to be annoying with an alert, so something like using the table footer would also work
      dispatch_async(dispatch_get_main_queue(), {
        let alertVC: UIAlertController = UIAlertController(title: "No Results", message: "please try again", preferredStyle: .Alert)
        let action: UIAlertAction = UIAlertAction(title: "OK", style: .Default, handler: nil)
        alertVC.addAction(action)
        self.presentViewController(alertVC, animated: true, completion: nil)
      })
      return
    } else if movies.count == 0 {
      // Empty page, nothing new to show so do nothing
      return
    } else if page == 1 {
      // Start over, new request!
      self.moviesDatasource.removeAll()
    }
    self.moviesDatasource.appendContentsOf(movies)
    
    // Always reload the view on the main thread
    dispatch_async(dispatch_get_main_queue(), {
      self.tableView.reloadData()
    })
  }
  
  func receivedMovieDetails(movie: Movie) {
    // Always make changes to the view heiarchy on the main thread
    dispatch_async(dispatch_get_main_queue(), {
      let detailsVC: MovieDetailsViewController = UIStoryboard(name: "Main", bundle: NSBundle.mainBundle()).instantiateViewControllerWithIdentifier("MovieDetailsViewController") as! MovieDetailsViewController
      detailsVC.movie = movie
      self.navigationController?.pushViewController(detailsVC, animated: true)
    })
  }
  
  func requestFailedWithError(error: NSError) {
    // An error happened, let the user know!
    // Ideally we would want user friendly messages, with some sort of next-step for the user to try
    dispatch_async(dispatch_get_main_queue(), {
      let alertVC: UIAlertController = UIAlertController(title: "Network Fail", message: error.description, preferredStyle: .Alert)
      let action: UIAlertAction = UIAlertAction(title: "OK", style: .Default, handler: nil)
      alertVC.addAction(action)
      self.presentViewController(alertVC, animated: true, completion: nil)
    })
  }
}